File: C#.Net/JDI/Web/JDI Web/Selenium/Elements/Complex/Menu.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using System.Threading;
using Epam.JDI.Core.Interfaces.Complex;
using JDI_Commons;
using OpenQA.Selenium;
using OpenQA.Selenium.Interactions;
using static Epam.JDI.Core.Settings.JDISettings;

namespace JDI_Web.Selenium.Elements.Complex
{
    public class Menu : Menu<IConvertible>, IMenu { }
    public class Menu<TEnum> : Selector<TEnum>, IMenu<TEnum>
        where TEnum : IConvertible
    {
        public List<By> MenuLevelsLocators = new List<By>();
        public string Separator = "\\>";
        public Menu<TEnum> UseSeparator(string separator)
        {
            Separator = separator;
            return this;
        }

        public Menu()
        {
            SelectNameAction = (m, name) => ChooseItemAction(this, new [] {name}, (w, el) => el.Click());
        }

        public Menu(By optionsNamesLocatorTemplate, List<IWebElement> webElements = null) 
            : base(optionsNamesLocatorTemplate, webElements)
        {
            MenuLevelsLocators.Add(optionsNamesLocatorTemplate);
        }

        public Menu(By optionsNamesLocatorTemplate, By allOptionsNamesLocator) 
            : base(optionsNamesLocatorTemplate, allOptionsNamesLocator) {
            MenuLevelsLocators.Add(optionsNamesLocatorTemplate);
        }
        public Menu(List<By> menuLevelsLocators)
        {
            MenuLevelsLocators = menuLevelsLocators;
        }

        public Action<IWebDriver, IWebElement> HoverAction = (driver, el) =>
        {
            var action = new Actions(driver);
            action.MoveToElement(el).ClickAndHold().Build().Perform();
        };

        public void Hover(params string[] names)
        {
            if (names == null || names.Length == 0)
                return;
            Actions.Hover(names.Print(Separator), (w, n) =>
            {
                var m = (Menu<TEnum>) w;
                m.ChooseItemAction(m, names, HoverAction);
            });
        }

        public int Delay = 200;
        
        private IList<string> SplitToList(string[] str, string separator)
        {
            return (str.Length == 1
                ? Regex.Split(str[0], separator)
                : str).ToList();
        }
        
        private void Select(By locator, string name)
        {
            new Selector(locator, element:this).Select(name);
        }

        protected Action<Menu<TEnum>, string[]> HoverAndClickAction = (m, names) =>
        {
            if (names == null || names.Length == 0)
                return;
            var split = m.SplitToList(names, m.Separator);
            if (split.Count == 1)
            {
                if (m.Delay > 0) Thread.Sleep(m.Delay);
                m.Select(m.Locator, split[0]);
                return;
            }
            if (split.Count > m.MenuLevelsLocators.Count)
                throw Exception($"Can't hover and click on element ({m}) by value: {names.Print(m.Separator)}. Amount of locators ({m.MenuLevelsLocators.Count}) less than select path length ({split.Count})");
            m.Hover(split.ListCopy(to: -1).ToArray());
            var lastIndex = split.Count - 1;
            if (m.Delay > 0) Thread.Sleep(m.Delay);
            m.Select(m.MenuLevelsLocators[lastIndex], split[lastIndex]);
        };
        public void Hover(TEnum name)
        {
            Hover(name.ToString());
        }

        public void Select(params string[] names)
        {
            HoverAndClick(names);
        }
        public void HoverAndClick(params string[] names)
        {
            Actions.Select(names.Print(Separator), (m, n) => HoverAndClickAction(this, names));
        }
        public void HoverAndClick(TEnum name)
        {
            HoverAndClick(name.ToString());
        }
        public void HoverAndSelect(params string[] name)
        {
            HoverAndClick(name);
        }
        public void HoverAndSelect(TEnum name)
        {
            HoverAndSelect(name.ToString());
        }

        protected Action<Menu<TEnum>, string[], Action<IWebDriver, IWebElement>> ChooseItemAction =
            (m, names, action) =>
            {
                var nodes = m.SplitToList(names, m.Separator);
                if (m.MenuLevelsLocators.Count == 0 && m.HasLocator)
                    m.MenuLevelsLocators.Add(m.Locator);
                if (m.MenuLevelsLocators.Count < nodes.Count) return;
                for (var i = 0; i < nodes.Count; i++)
                {
                    var value = nodes[i];
                    var elements = new Selector(m.MenuLevelsLocators[i], element: m).Elements;
                    var element = elements.FirstOrDefault(el => el.Text.Equals(value));
                    if (element == null)
                        throw Exception("Can't choose element:" + value);
                    if (m.Delay > 0) Thread.Sleep(m.Delay);
                    action(m.WebDriver, element);
                }
            };

        public void SetUp(List<By> menuLevelsLocators)
        {
            MenuLevelsLocators = menuLevelsLocators;
        }
    }
}

File: Java/JDI/jdi-uitest-mobile/src/main/java/com/epam/jdi/uitests/mobile/appium/elements/complex/Menu.java
package com.epam.jdi.uitests.mobile.appium.elements.complex;
/*
 * Copyright 2004-2016 EPAM Systems
 *
 * This file is part of JDI project.
 *
 * JDI is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * JDI is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with JDI. If not, see <http://www.gnu.org/licenses/>.
 */


import com.epam.jdi.uitests.core.interfaces.complex.IMenu;
import com.epam.jdi.uitests.mobile.appium.elements.base.Element;
import org.openqa.selenium.By;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

import static com.epam.commons.EnumUtils.getEnumValue;
import static com.epam.commons.LinqUtils.first;
import static com.epam.commons.PrintUtils.print;
import static com.epam.jdi.uitests.core.settings.JDISettings.exception;
import static java.util.Arrays.copyOfRange;

/**
 * RadioButtons control implementation
 *
 * @author Alexeenko Yan
 */
public class Menu<TEnum extends Enum> extends Selector<TEnum> implements IMenu<TEnum> {
    private List<By> menuLevelsLocators = new ArrayList<>();
    private String delimeter = "\\|";
    public <T extends IMenu<TEnum>> T useSeparator(String separator) {
        this.delimeter = separator;
        return (T)this;
    }

    public Menu() { }

    public Menu(By optionsNamesLocatorTemplate) {
        super(optionsNamesLocatorTemplate);
    }

    public Menu(By optionsNamesLocatorTemplate, By allOptionsNamesLocator) {
        super(optionsNamesLocatorTemplate, allOptionsNamesLocator);
        menuLevelsLocators.add(optionsNamesLocatorTemplate);
    }
    public Menu(List<By> menuLevelsLocators) {
        this.menuLevelsLocators = menuLevelsLocators;
    }

    protected void hoverAction(String... name) {
        chooseItemAction(name, el -> {
            Actions action = new Actions(getDriver());
            action.moveToElement(el).clickAndHold().build().perform();
        });
    }

    public final void hover(String... names) {
        if (names == null || names.length == 0)
            return;
        actions.hover(print(names, delimeter), n ->
                hoverAction(SplitToList(names)));
    }

    private String[] SplitToList(String[] str)
    {
        return str.length == 1
                ? str[0].split(delimeter)
                : str;
    }

    protected void hoverAndClickAction(String... names) {
        if (names == null || names.length == 0)
            return;
        String[] split = SplitToList(names);
        if (split.length > menuLevelsLocators.size())
            throw exception("Can't hover and click on element (%s) by value: %s. Amount of locators (%s) less than select path length (%s)",
                    this, print(names, delimeter), menuLevelsLocators.size(), split.length);
        if (split.length > 1)
            hover(copyOfRange(split, 0, split.length-1));
        int lastIndex = split.length - 1;
        Selector selector = new Selector<>(menuLevelsLocators.get(lastIndex));
        selector.setParent(getParent());
        selector.select(split[lastIndex]);
    }
    public final void hover(TEnum name) {
        hover(getEnumValue(name));
    }
    public final void hoverAndClick(String... names) { actions.select(print(names, delimeter), this::hoverAndClickAction); }
    public final void hoverAndClick(TEnum name) {
        hoverAndClick(getEnumValue(name));
    }
    public final void hoverAndSelect(String... names) {
        hoverAndClick(names);
    }
    public final void hoverAndSelect(TEnum name) {
        hoverAndClick(getEnumValue(name));
    }

    protected void selectAction(String... names) {
        hoverAndClick(names);
    }

    protected void chooseItemAction(String[] names,  Consumer<WebElement> action) {
        String[] nodes = SplitToList(names);
        if (menuLevelsLocators.size() == 0 && hasLocator())
            menuLevelsLocators.add(getLocator());
        if (menuLevelsLocators.size() < nodes.length) return;
        for(int i = 0; i < nodes.length; i++) {
            String value = nodes[i];
            List<WebElement> elements = new Selector<>(menuLevelsLocators.get(i)).getElements();
            WebElement element = first(elements, el -> el.getText().equals(value));
            new Element(element).invoker.processDemoMode();
            action.accept(element);
        }
    }
    public void select(String... names) {
        select(print(names, delimeter));
    }

    public void setUp(List<By> menuLevelsLocators) {
        this.menuLevelsLocators = menuLevelsLocators;
    }
}